cpp/src/arrow/util/io_util.cc

```diff
@@ -203,16 +203,26 @@ std::string ErrnoMessage(int errnum) { return std::strerror(errnum); }
 
 #if _WIN32
 std::string WinErrorMessage(int errnum) {
-  char buf[1024];
-  auto nchars = FormatMessageA(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,
-                               NULL, errnum, 0, buf, sizeof(buf), NULL);
-  if (nchars == 0) {
+#define MAX_N_CHARACTERS 1024
+#define MAX_N_UTF8_BYTES 4096
+  WCHAR utf16_message[MAX_N_CHARACTERS];
+  auto n_utf16_chars =
+      FormatMessageW(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS, NULL,
+                     errnum, 0, utf16_message, sizeof(utf16_message), NULL);
+  if (n_utf16_chars == 0) {
     // Fallback
     std::stringstream ss;
     ss << "Windows error #" << errnum;
     return ss.str();
   }
-  return std::string(buf, nchars);
+  const UINT code_page = CP_UTF8;
+  const DWORD flags = 0;
+  char utf8_message[MAX_N_UTF8_BYTES];
+  auto n_utf8_bytes = WideCharToMultiByte(code_page, flags, utf16_message, n_utf16_chars,
```

Comment:
You can use `WideStringToUTF8`.

Reply:
Note that `WideCharToMultiByte` is converting to the current code page, but we want to convert to utf8 instead.

Reply:
Who does provide `WideStringToUTF8`?
It seems that Windows doesn't provide it.

The code uses `CP_UTF8` not `CP_ACP` here. So this converts to UTF-8 not the current code page.

Reply:
`WideStringToUTF8` is one of our helper functions.

Reply:
Ah, thank. I could find it.

Reply:
I've changed to use `WideStringToUTF8()`.